### PR TITLE
Implement of Zero-Shot CLIP Classifier

### PR DESCRIPTION
**Pull request by Coobiw**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

I implement a version of zero-shot CLIP. With evaluation on CIFAR100 and ImageNet1k, it has a small gap with the official one(OpenAI's). After adding sub-prompt(whose number is 8, 1/10 of full prompt-engineering previously), the performance becomes better.

## Modification

I've added the file in mmpretrain.models.multimodal.clip_zs and the config files in configs.clip_zs. Just run:

```
python tools/test.py configs/clip/clip_vit-base-p16_zeroshot-cls_cifar100.py clip-vit-b-p16_converted.pth
```

Get the checkpoint
```
pip install ftfy regex tqdm
pip install git+https://github.com/openai/CLIP.git

import clip
model,_ = clip.load('ViT-B/16')
ckpt = model.state_dict()
torch.save(ckpt,'clip-vit-b-p16.pth'
```

then run the convert script `tools/model_converters/openai-clip_to_mmpretrain-clip.py`:
```
python tools/model_converters/openai-clip_to_mmpretrain-clip.py clip-vit-b-p16.pth clip-vit-b-p16_converted.pth
```

I've uploaded the conveted weights on [clip-vit-b-p16_converted.pth](https://github.com/Coobiw/modelhub/releases/download/mmpretrain-clip/clip-vit-b-p16_converted.pth) and [clip-vit-l-p14_converted.pth](https://github.com/Coobiw/modelhub/releases/download/mmpretrain-clip/clip-vit-l-p14_converted.pth)

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

```
# ViT-B/16 for CIFAR100
python tools/test.py configs/clip/clip_vit-base-p16_zeroshot-cls_cifar100.py clip-vit-b-p16_converted.pth

# ViT-L/14 for CIFAR100
python tools/test.py configs/clip/clip_vit-large-p14_zeroshot-cls_cifar100.py clip-vit-l-p14_converted.pth

# ViT-B/16 for IN1k
python tools/test.py configs/clip/clip_vit-base-p16_zeroshot-cls_in1k.py clip-vit-b-p16_converted.pth

# ViT-L/14 for IN1k
python tools/test.py configs/clip/clip_vit-large-p14_zeroshot-cls_in1k.py clip-vit-l-p14_converted.pth
```
The result is shown in following format:
image

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpretrain/models/multimodal/clip_zs/clip.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward function."""
+        orig_type = x.dtype
+        ret = super().forward(x.type(torch.float32))
+        return ret.type(orig_type)
+
+
+@MODELS.register_module()
+class CLIP_zs(BaseModel):
+    """The implementation of `ChineseCLIP <https://arxiv.org/abs/2211.01335>`_.
```
Comment (fangyixiao18) [inline mmpretrain/models/multimodal/clip_zs/clip.py:43]: update docstrings to CLIP and CLIP's link

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: mmpretrain/models/multimodal/clip_zs/clip.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+        for i, tokens in enumerate(all_tokens):
+            assert len(tokens) <= self.context_length
+            result[i, :len(tokens)] = torch.tensor(tokens)
+
+        return result
+
+
+@MODELS.register_module()
+class CLIP_zs(CLIP):
```
Comment (fangyixiao18) [inline mmpretrain/models/multimodal/clip_zs/clip.py:274]: Rename to CLIPZeroShot, please also check other parts

File: configs/clip_zs/clip-vit-base-patch16_cifar100.py
Comment (fangyixiao18) [inline configs/clip_zs/clip-vit-base-patch16_cifar100.py:1]: move all these 4 configs to `configs/clip/` folder, and rename them to format like `clip_vit-base-p16_zeroshot-cls_cifar100.py`

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

**Review by fangyixiao18 — APPROVED** [review]

**Comment by npurson** [comment]
Appreciating your implementation! It really helped me out a lot!

BTW, have you ever considered adding the usage instructions to the README within the clip configs? I struggled to find how to use this feature until I dug through the PRs.

**Pull request merged**